lib/galaxy/datatypes/isa.py

```diff
@@ -0,0 +1,520 @@
+        """Create the HTML table used for displaying peek, from the peek text found by set_peek() method."""
+
+        out = ['<table cellspacing="0" cellpadding="3">']
+        try:
+            if not dataset.peek:
+                dataset.set_peek()
+            json_data = json.loads(dataset.peek)
+            for line in json_data["data"]:
+                line = line.strip()
+                if not line:
+                    continue
+                out.append('<tr><td>%s</td></tr>' % escape(util.unicodify(line, 'utf-8')))
+            out.append('</table>')
+            out = "".join(out)
+        except Exception as exc:
+            out = "Can't create peek %s" % str(exc)
+        return out
+
+    # Generate primary file {{{2
+    ################################################################
+
+    def generate_primary_file(self, dataset=None):
+        """Generate the primary file. It is an HTML file containing description of the composite dataset
+           as well as a list of the composite files that it contains."""
+
+        if dataset:
+            rval = ['<html><head><title>ISA Dataset </title></head><p/>']
+            if hasattr(dataset, "extra_files_path"):
+                rval.append('<div>ISA Dataset composed of the following files:<p/><ul>')
+                for cmp_file in os.listdir(dataset.extra_files_path):
+                    opt_text = ''
+                    rval.append('<li><a href="%s" type="text/plain">%s</a>%s</li>' % (cmp_file, cmp_file, opt_text))
+                rval.append('</ul></div></html>')
+            else:
+                rval.append('<div>ISA Dataset is empty!<p/><ul>')
+            return "\n".join(rval)
+        return "<div>No dataset available</div>"
+
+    # Dataset content needs grooming {{{2
+    ################################################################
+
+    def dataset_content_needs_grooming(self, file_name):
+        """This function is called on an output dataset file after the content is initially generated."""
+        return os.path.basename(file_name) == ISA_ARCHIVE_NAME
+
+    # Groom dataset content {{{2
+    ################################################################
+
+    def groom_dataset_content(self, file_name):
+        """This method is called by Galaxy to extract files contained in a composite data type."""
+        # XXX Is the right place to extract files? Should this step not be a cleaning step instead?
+        # Could extracting be done earlier and composite files declared as files contained inside the archive
+        # instead of the archive itself?
+
+        # extract basename and folder of the current file whose content has to be groomed
+        basename = os.path.basename(file_name)
+        output_path = os.path.dirname(file_name)
+        # extract archive if the file corresponds to the ISA archive
+        if basename == ISA_ARCHIVE_NAME:
+            # perform extraction
+            with open(file_name, 'rb') as stream:
```

Comment:
Does 

```
from galaxy.util.compression_utils import CompressedFile
CompressedFile(file_name).extract(output_path)
```

work here? It'd simply the implementation a lot I think if you did that. Could then get rid of _extract_* helpers, maybe that regex at the top, etc....